src/main/java/co/rsk/federate/watcher/FederationWatcher.java

```diff
@@ -0,0 +1,124 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(FederationWatcher.class);
+
+    private final Ethereum rsk;
+
+    private FederationProvider federationProvider;
+    private FederationWatcherListener federationWatcherListener; 
+
+    private Federation activeFederation;
+    private Federation retiringFederation;
+
+    /**
+     * Constructs a new {@code FederationWatcher} with the specified RSK client.
+     *
+     * @param rsk the Ethereum client used to listen for new blocks on the RSK blockchain
+     */
+    public FederationWatcher(Ethereum rsk) {
+        this.rsk = rsk;
+    }
+
+    /**
+     * Starts the {@code FederationWatcher} by setting the {@code FederationProvider}
+     * and {@code FederationWatcherListener}, and begins listening for new blocks on
+     * the RSK blockchain.
+     *
+     * @param federationProvider        the provider used to obtain federation information (active and retiring federations)
+     * @param federationWatcherListener the listener that will be notified when the federations change
+     */
+    public void start(FederationProvider federationProvider, FederationWatcherListener federationWatcherListener) {
+        this.federationProvider = federationProvider;
+        this.federationWatcherListener = federationWatcherListener;
+     
+        rsk.addListener(new EthereumListenerAdapter() {
+            @Override
+            public void onBestBlock(org.ethereum.core.Block block, List<TransactionReceipt> receipts) {
+                // Updating state only when the best block changes still "works",
+                // since we're interested in finding out only when the active or retiring federation(s) changed.
+                // If there was a side chain in which any of these changed in, say, block 4500, but
+                // that side chain became main chain in block 7800, it would still be ok to
+                // start monitoring the new federation(s) on that block.
+                // The main reasons for this is that RSK nodes would have never reported the active or
+                // retiring federation(s) being different *before* the best chain change. Therefore,
+                // there should be no new Bitcoin transactions directed to these new addresses
+                // until this change effectively becomes a part of RSK's "reality".
+                // The case in which we go back and forth to and from a sidechain in which the
+                // federation effectively changed is still to be explored deeply, but the same reasoning
+                // should apply since going back and forth would trigger two federation changes.
+                // A client trying to send bitcoins to the new federation without waiting
+                // a good number of confirmations would be, essentially, "playing with fire".
+                logger.info("[updateState] New best block, updating state");
+                updateState();
+            }
+        });
```

Comment:
not a big fan of having this block of code inside the `start` method. I think before was easier to follow

Reply:
The idea is to avoid nested inner clases if possible. This might seem big because of the comment. I have reorganized the comment to be more readable.